twine/auth.py

```diff
@@ -0,0 +1,92 @@
+import warnings
+import getpass
+import functools
+import typing
+from typing import Optional, Callable
+
+import keyring
+
+from . import utils
+from . import exceptions
+
+
+class CredentialInput(typing.NamedTuple):
+    username: typing.Optional[str]
+    password: typing.Optional[str]
+
+    @classmethod
+    def new(cls, username=None, password=None):
+        return cls(username, password)
+
+
+class Resolver:
+    def __init__(self, config: utils.RepositoryConfig, input: CredentialInput):
+        self.config = config
+        self.input = input
+
+    @classmethod
+    def choose(cls, interactive):
+        return cls if interactive else Private
+
+    @property  # type: ignore  # https://github.com/python/mypy/issues/1362
+    @functools.lru_cache()
+    def username(self) -> Optional[str]:
+        return utils.get_userpass_value(
+            self.input.username,
+            self.config,
+            key='username',
+            prompt_strategy=self.username_from_keyring_or_prompt,
+        )
+
+    @property  # type: ignore  # https://github.com/python/mypy/issues/1362
+    @functools.lru_cache()
+    def password(self) -> Optional[str]:
+        return utils.get_userpass_value(
+            self.input.password,
+            self.config,
+            key='password',
+            prompt_strategy=self.password_from_keyring_or_prompt,
+        )
+
+    @property
+    def system(self) -> Optional[str]:
+        return self.config['repository']
+
+    def get_username_from_keyring(self):
+        try:
+            creds = keyring.get_credential(self.system, None)
+            if creds:
+                return creds.username
+        except AttributeError:
+            # To support keyring prior to 15.2
+            pass
+        except Exception as exc:
+            warnings.warn(str(exc))
+
+    def get_password_from_keyring(self) -> Optional[str]:
+        try:
+            return keyring.get_password(self.system, self.username)
+        except Exception as exc:
+            warnings.warn(str(exc))
+        return None  # TODO: mypy shouldn't require this
+        return None  # any more than it should require this
```

Comment:
Why are there two `return`s here? I'm surprised this passed linting.

Reply:
Yeah, that's me being a little pedantic in reaction to mypy being unnecessarily pedantic. Mypy requires the first `return None`, which in my mind is redundant lint. I learned over 20 years ago that a Python function that reaches the end of the block without returning anything has an implicit `return None`, but now mypy is violating that language construct by requiring an explicit return.